compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -863,27 +869,59 @@ impl<'tcx> ThirBuildCx<'tcx> {
                     .iter()
                     .any(|attr| attr.has_name(sym::loop_match));
                 if is_loop_match {
-                    assert!(body.stmts.is_empty());
-                    let hir::ExprKind::Assign(state, block_expr, _) = body.expr.unwrap().kind
-                    else {
-                        panic!();
+                    if let Some(first) = body.stmts.first() {
+                        let span = first.span.to(body.stmts.last().unwrap().span);
+                        self.tcx.dcx().emit_fatal(LoopMatchBadStatements { span })
+                    }
+
+                    let Some(loop_body_expr) = body.expr else {
+                        self.tcx.dcx().emit_fatal(LoopMatchMissingAssignment { span: body.span })
+                    };
+
+                    let hir::ExprKind::Assign(state, rhs_expr, _) = loop_body_expr.kind else {
+                        let span = loop_body_expr.span;
+                        self.tcx.dcx().emit_fatal(LoopMatchMissingAssignment { span })
+                    };
+
+                    let hir::ExprKind::Block(block_body, _) = rhs_expr.kind else {
+                        let span = rhs_expr.span;
+                        self.tcx.dcx().emit_fatal(LoopMatchBadRhs { span })
                     };
-                    let hir::ExprKind::Block(block_body, _) = block_expr.kind else {
-                        panic!();
+
+                    if let Some(first) = block_body.stmts.first() {
+                        let span = first.span.to(block_body.stmts.last().unwrap().span);
+                        self.tcx.dcx().emit_fatal(LoopMatchBadStatements { span })
+                    }
+
+                    let Some(block_body_expr) = block_body.expr else {
+                        self.tcx.dcx().emit_fatal(LoopMatchBadRhs { span: block_body.span })
                     };
-                    assert!(block_body.stmts.is_empty());
-                    let hir::ExprKind::Match(discr, arms, match_source) =
-                        block_body.expr.unwrap().kind
+
+                    let hir::ExprKind::Match(scrutinee, arms, match_source) = block_body_expr.kind
                     else {
-                        panic!();
+                        let span = block_body_expr.span;
+                        self.tcx.dcx().emit_fatal(LoopMatchBadRhs { span })
+                    };
+
+                    fn local(expr: &rustc_hir::Expr<'_>) -> Option<hir::HirId> {
+                        if let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = expr.kind {
+                            if let Res::Local(hir_id) = path.res {
+                                return Some(hir_id);
+                            }
+                        }
+
+                        None
+                    }
```

Comment:
ideally this would be a method somewhere, but I'm not sure where that should be

Reply:
Not sure either. Can be left for later.